File: apps/subgraph/src/log.ts
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Bytes, ethereum } from "@graphprotocol/graph-ts";
import { Log } from "../generated/schema";
import {
  BEFORE_EXECUTION_EVENT_HASH,
  USER_OPERATION_EVENT_HASH,
} from "./const";

export function handleUserOperationLogs(
  userOpHash: Bytes,
  eventTransaction: ethereum.Transaction,
  eventReceipt: ethereum.TransactionReceipt | null,
): Log[] {
  const logs = new Array<Log>();

  if (eventReceipt) {
    // Flag to store the log in relation to the user operation
    let flag = false;

    // Iterate through the receipt logs in reverse order
    // This is because the `UserOperationEvent` event is emitted after the logs are emitted
    // Break until the `BeforeExecution` event is emitted
    for (let i = eventReceipt.logs.length - 1; i >= 0; i--) {
      // Load the Log entity
      // Get the index from the log in reverse order
      const log = Log.load(`${eventTransaction.hash.toHexString()}-${i}`);

      // If the Log entity doesn't exist, break;
      if (log == null) {
        break;
      }

      // Get the topic from the log
      const topic = eventReceipt.logs[i].topics[0];

      // If the topic is an `UserOperationEvent` topic, get the user operation hash
      if (topic.toHexString() === USER_OPERATION_EVENT_HASH) {
        // Get the log user operation hash from the log (the first topic is the event hash)
        const logUserOpHash = eventReceipt.logs[i].topics[1];
        // If the log user operation hash is equal to the event user operation hash, set the flag to true
        if (logUserOpHash === userOpHash) {
          flag = true;
        } else {
          // If the log user operation hash is not equal to the event user operation hash, set the flag to false
          // Could be break, since we handle the logs for a `UserOperationEvent` event
          flag = false;
        }
      }

      // If the topic is an `BeforeExecution` topic, break
      if (topic.toHexString() === BEFORE_EXECUTION_EVENT_HASH) {
        flag = false;
        break;
      }

      // If the flag is true, set the user operation hash to the log
      if (flag) {
        log.userOperation = userOpHash;
        log.save();
      }
    }
  }

  return logs;
}

File: crates/db/src/models/log.rs
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![allow(clippy::unwrap_used)]

use alloy::{
    primitives::{Log as AlloyLog, B256},
    rpc::types::Log,
};
use eyre::{eyre, Result};
use lightdotso_prisma::log;
use revm::primitives::LogData;
use serde::{Deserialize, Serialize};

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DbLog {
    pub log: Log,
    pub topics: Vec<B256>,
}

impl TryFrom<log::Data> for DbLog {
    type Error = eyre::Report;

    fn try_from(log: log::Data) -> Result<Self, Self::Error> {
        let topics = log.clone().topics.unwrap().into_iter().collect::<Vec<_>>();

        // Parses and obtains the topic string and its index.
        let mut log_topic_ids = topics
            .into_iter()
            .map(|t| {
                let split = t.id.split('-').collect::<Vec<_>>();

                // Ensure there are enough parts
                if split.len() < 2 {
                    return Err(eyre!("Unexpected topic format. Expected '<id>-<index>'."));
                }

                // Parse the index
                let index =
                    split[1].parse::<usize>().map_err(|_| eyre!("Invalid number in topic id"))?;

                // Parse log_topic_id
                let log_topic_id =
                    split[0].parse::<B256>().map_err(|_| eyre!("Error while parsing log topic"))?;

                Ok((index, log_topic_id))
            })
            .collect::<Result<Vec<(usize, B256)>>>()?;

        // Sorts topics by their indexes.
        log_topic_ids.sort_by_key(|(index, _)| *index);

        // If only the sorted log_topic_ids are needed, map again to remove the index.
        let log_topics =
            log_topic_ids.into_iter().map(|(_, log_topic_id)| log_topic_id).collect::<Vec<B256>>();

        let alloy_log =
            AlloyLog::new(log.address.parse().unwrap(), log_topics.clone(), log.data.into())
                .unwrap();

        let log = Log::<LogData> {
            inner: alloy_log,
            block_hash: log.block_hash.map(|bh| bh.parse().unwrap()),
            block_number: log.block_number.map(|bn| (bn as u64)),
            block_timestamp: None,
            transaction_hash: log.transaction_hash.map(|th| th.parse().unwrap()),
            transaction_index: log.transaction_index.map(|ti| (ti as u64)),
            log_index: log.log_index.map(|li| (li as u64)),
            removed: false,
        };

        Ok(Self { log: log.clone(), topics: log_topics })
    }
}
